File: argus/packs/chunk.py
"""Heading-aware chunking for prose/documentation packs.

Code chunks at function boundaries; prose has no such structure, so it
chunks at Markdown headings instead. The critical property is not the
splitting itself but what gets embedded: a section body on its own
("A string indicating how to handle a redirect response.") embeds to
almost nothing useful, because it could be about anything. Prepending the
heading trail ("fetch() > Parameters > options > redirect") makes the
chunk self-locating. See ``embed_text`` -- that prepending is the entire
point of this module.

Nothing here may import ``argus.store.queries`` or open ``index.db``;
packs are a second, public corpus entirely separate from the private index.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

_HEADING_RE = re.compile(r'^(#{1,6})[ \t]+(.+?)[ \t]*#*[ \t]*$')
_FENCE_RE = re.compile(r'^(```+|~~~+)')


def _closes_fence(fence_match: re.Match[str], opener: tuple[str, int]) -> bool:
    """A closing fence must use the same character as the opener and be at
    least as long (CommonMark). A different character, or a shorter run of
    the same character, is not a closer -- it's just fenced content (e.g. a
    ``~~~`` line shown inside a backtick-fenced block of documentation)."""
    marker = fence_match.group(1)
    char, min_length = opener
    return marker[0] == char and len(marker) >= min_length


_SLUG_STRIP_RE = re.compile(r'[^\w\s-]')
_SLUG_WHITESPACE_RE = re.compile(r'[\s_]+')


@dataclass
class Chunk:
    """One retrievable unit of a documentation pack.

    ``heading_path`` is the full trail from the document's top-level
    heading down to whichever heading directly governs ``body`` (for
    example ``"fetch() > Parameters > options"``); it is ``""`` only for
    content that appears before any heading at all. ``anchor`` is the slug
    of the immediate (deepest) heading, for deep-linking back to the
    source. ``start_line`` is the 1-indexed source line the body begins
    at.
    """

    heading_path: str
    anchor: str | None
    start_line: int
    body: str


def _slugify(title: str) -> str:
    slug = title.strip().lower()
    slug = _SLUG_STRIP_RE.sub("", slug)
    slug = _SLUG_WHITESPACE_RE.sub("-", slug)
    return slug.strip("-")


def _dedupe_anchor(base: str, seen: dict[str, int]) -> str:
    """Make ``base`` unique within a document, the conventional way: the
    first heading to produce a given slug keeps it bare, later ones get a
    numbered suffix (``abortcontroller``, ``abortcontroller-2``, ...).

    Without this, two headings that slug identically -- an H1
    ``AbortController`` and an H3 ``AbortController()`` are a real,
    MDN-shaped example -- would deep-link to the same anchor and silently
    strand the reader at the wrong one.
    """
    count = seen.get(base, 0) + 1
    seen[base] = count
    return base if count == 1 else f"{base}-{count}"


@dataclass
class _Section:
    heading_path: str
    anchor: str | None
    start_line: int
    lines: list[str]


def _split_into_sections(text: str) -> list[_Section]:
    """Walk the document once, maintaining a heading stack.

    A heading at level *n* pops every stack entry at level >= n before
    pushing itself -- get this backwards (level > n) and same-level
    siblings nest into each other; get the whole rule wrong and deeper
    headings leak into their uncles (``options > Return value``).

    Fenced code blocks are tracked here too, purely so a ``#`` inside one
    is never mistaken for a heading.
    """
    lines = text.splitlines()
    stack: list[tuple[int, str]] = []
    sections: list[_Section] = []
    current_lines: list[str] = []
    current_start = 1
    fence_opener: tuple[str, int] | None = None
    seen_anchors: dict[str, int] = {}

    def flush() -> None:
        heading_path = " > ".join(title for _, title in stack)
        if stack:
            anchor = _dedupe_anchor(_slugify(stack[-1][1]), seen_anchors)
        else:
            anchor = None
        sections.append(_Section(heading_path, anchor, current_start,
                                  list(current_lines)))

    for lineno, line in enumerate(lines, start=1):
        stripped = line.strip()
        fence_match = _FENCE_RE.match(stripped)

        if fence_opener is not None:
            current_lines.append(line)
            if fence_match and _closes_fence(fence_match, fence_opener):
                fence_opener = None
            continue

        if fence_match:
            fence_opener = (fence_match.group(1)[0], len(fence_match.group(1)))
            current_lines.append(line)
            continue

        heading_match = _HEADING_RE.match(line)
        if heading_match:
            flush()
            level = len(heading_match.group(1))
            title = heading_match.group(2).strip()
            while stack and stack[-1][0] >= level:
                stack.pop()
            stack.append((level, title))
            current_lines = []
            current_start = lineno + 1
            continue

        current_lines.append(line)

    flush()
    return [s for s in sections if "\n".join(s.lines).strip()]


@dataclass
class _Block:
    start_line: int
    text: str
    is_fence: bool


def _split_into_blocks(lines: list[str], start_line: int) -> list[_Block]:
    """Split section lines into paragraphs, keeping each fenced code block
    (open marker to close marker, blank lines and all) as one atomic block
    that later splitting must never break open."""
    blocks: list[_Block] = []
    cur: list[str] = []
    cur_start = start_line
    fence_opener: tuple[str, int] | None = None

    for offset, line in enumerate(lines):
        lineno = start_line + offset
        stripped = line.strip()
        fence_match = _FENCE_RE.match(stripped)

        if fence_opener is not None:
            cur.append(line)
            if fence_match and _closes_fence(fence_match, fence_opener):
                blocks.append(_Block(cur_start, "\n".join(cur), True))
                cur = []
                fence_opener = None
            continue

        if fence_match:
            if cur:
                blocks.append(_Block(cur_start, "\n".join(cur), False))
                cur = []
            cur_start = lineno
            cur = [line]
            fence_opener = (fence_match.group(1)[0], len(fence_match.group(1)))
            continue

        if stripped == "":
            if cur:
                blocks.append(_Block(cur_start, "\n".join(cur), False))
                cur = []
            continue

        if not cur:
            cur_start = lineno
        cur.append(line)

    if cur:
        blocks.append(_Block(cur_start, "\n".join(cur), fence_opener is not None))

    return blocks


def _wrap_by_chars(text: str, max_chars: int) -> list[str]:
    """Split a single oversized paragraph at word boundaries. Used only for
    non-fence blocks, since a fence must never be broken regardless of size."""
    words = text.split()
    if not words:
        return []

    pieces: list[str] = []
    cur: list[str] = []
    cur_len = 0
    for word in words:
        extra = len(word) + (1 if cur else 0)
        if cur and cur_len + extra > max_chars:
            pieces.append(" ".join(cur))
            cur = [word]
            cur_len = len(word)
        else:
            cur.append(word)
            cur_len += extra
    if cur:
        pieces.append(" ".join(cur))
    return pieces


def _pack_units(units: list[tuple[int, str]], max_chars: int) -> list[tuple[int, str]]:
    """Greedily pack paragraph/fence units into chunks under max_chars.
    A single unit that is already oversized (an unsplittable fence, or a
    paragraph already at the word-wrap limit) is still emitted alone
    rather than dropped or forced together with a neighbor."""
    packed: list[tuple[int, str]] = []
    cur_texts: list[str] = []
    cur_start = 0
    cur_len = 0

    for start_line, text in units:
        extra = len(text) + (2 if cur_texts else 0)
        if cur_texts and cur_len + extra > max_chars:
            packed.append((cur_start, "\n\n".join(cur_texts)))
            cur_texts = [text]
            cur_start = start_line
            cur_len = len(text)
        else:
            if not cur_texts:
                cur_start = start_line
            cur_texts.append(text)
            cur_len += extra

    if cur_texts:
        packed.append((cur_start, "\n\n".join(cur_texts)))

    return packed


def chunk_markdown(text: str, *, max_chars: int = 1200) -> list[Chunk]:
    """Chunk Markdown prose at heading boundaries.

    Every chunk carries the full heading trail down to whichever heading
    governs it. Oversized sections are split at paragraph boundaries (and,
    if a single paragraph alone exceeds ``max_chars``, at word boundaries),
    but a fenced code block is never split -- a half-open fence would
    poison every downstream consumer of that chunk.
    """
    chunks: list[Chunk] = []

    for section in _split_into_sections(text):
        blocks = _split_into_blocks(section.lines, section.start_line)

        units: list[tuple[int, str]] = []
        for block in blocks:
            if block.is_fence or len(block.text) <= max_chars:
                units.append((block.start_line, block.text))
            else:
                units.extend(
                    (block.start_line, piece)
                    for piece in _wrap_by_chars(block.text, max_chars)
                )

        for start_line, body in _pack_units(units, max_chars):
            chunks.append(Chunk(
                heading_path=section.heading_path,
                anchor=section.anchor,
                start_line=start_line,
                body=body,
            ))

    return chunks


def embed_text(chunk: Chunk) -> str:
    """The string that actually gets embedded.

    This is the entire reason the module exists: prepend the heading
    trail so the chunk is self-locating. A pack built from
    ``chunk.body`` alone looks identical in every metric that isn't
    retrieval quality -- same chunk count, same sizes -- and simply
    returns worse answers forever.
    """
    if not chunk.heading_path:
        return chunk.body
    return f"{chunk.heading_path}\n{chunk.body}"

File: src/Argus/Packs/Chunk.cs
using System.Text.RegularExpressions;
using Argus.Util;

namespace Argus.Packs;

public sealed record Chunk(string HeadingPath, string? Anchor, int StartLine, string Body);

/// <summary>
/// Markdown (and reStructuredText, converted) into retrieval chunks:
/// split at headings, keep fenced blocks whole, pack
/// paragraphs up to a size, and carry the heading trail and a stable anchor.
/// </summary>
public static class Chunker
{
    const RegexOptions O = RegexOptions.CultureInvariant;
    static readonly Regex HeadingRe = new(@"^(#{1,6})[ \t]+(.+?)[ \t]*#*[ \t]*$", O);
    static readonly Regex FenceRe = new(@"^(```+|~~~+)", O);
    static readonly Regex PinnedAnchorRe = new(@"\s*\{/\*\s*(.*?)\s*\*/\}\s*$", O);
    static readonly Regex RstRuleRe = new(@"^([=\-~^""'`#*+:.,_])\1{2,}\s*$", O);
    static readonly Regex RstRoleRe = new(@":[\w:+-]+:`([^`]*)`", O);
    static readonly Regex RstLiteralRe = new(@"``([^`]*)``", O);
    static readonly Regex RstInterpretedRe = new(@"`([^`]*)`", O);
    static readonly Regex RstEmphasisRe = new(@"\*{1,2}([^*]+)\*{1,2}", O);
    static readonly Regex SlugStripRe = new(@"[^\w\s-]", O);
    static readonly Regex SlugWhitespaceRe = new(@"[\s_]+", O);

    static bool ClosesFence(string marker, (char Char, int Min) opener) => marker[0] == opener.Char && marker.Length >= opener.Min;

    static string RoleText(Match m)
    {
        var inner = m.Groups[1].Value;
        if (inner.StartsWith('!')) return inner.Substring(1);
        if (inner.StartsWith('~')) return PyStr.AfterLast(inner.Substring(1), '.');
        return inner;
    }

    public static string StripRstInline(string text)
    {
        text = RstRoleRe.Replace(text, RoleText);
        text = RstLiteralRe.Replace(text, "$1");
        text = RstInterpretedRe.Replace(text, "$1");
        text = RstEmphasisRe.Replace(text, "$1");
        return PyStr.Strip(text);
    }

    /// <summary>Rewrite reST section titles as ATX headings, levels in order of first appearance.</summary>
    public static string RstToAtx(string text)
    {
        var lines = PyStr.SplitLines(text);
        var output = new List<string>(lines);
        var styles = new List<(char, bool)>();
        int LevelFor((char, bool) style)
        {
            if (!styles.Contains(style)) styles.Add(style);
            return Math.Min(styles.IndexOf(style) + 1, 6);
        }
        int index = 0;
        while (index < lines.Count)
        {
            var current = PyStr.Strip(lines[index]);
            var over = RstRuleRe.IsMatch(current);
            if (over && index + 2 < lines.Count)
            {
                var title = PyStr.Strip(lines[index + 1]);
                var below = PyStr.Strip(lines[index + 2]);
                if (title.Length > 0 && !RstRuleRe.IsMatch(title) && RstRuleRe.IsMatch(below) && below[0] == current[0])
                {
                    var marker = new string('#', LevelFor((current[0], true)));
                    output[index] = "";
                    output[index + 1] = $"{marker} {StripRstInline(title)}";
                    output[index + 2] = "";
                    index += 3;
                    continue;
                }
            }
            if (current.Length > 0 && !over && index + 1 < lines.Count)
            {
                var below = PyStr.Strip(lines[index + 1]);
                if (RstRuleRe.IsMatch(below) && PyStr.Len(below) >= PyStr.Len(current))
                {
                    var marker = new string('#', LevelFor((below[0], false)));
                    output[index] = $"{marker} {StripRstInline(current)}";
                    output[index + 1] = "";
                    index += 2;
                    continue;
                }
            }
            index++;
        }
        return string.Join("\n", output);
    }

    public static List<bool> FencedLineFlags(IReadOnlyList<string> lines)
    {
        var flags = new List<bool>();
        (char, int)? opener = null;
        foreach (var line in lines)
        {
            var m = FenceRe.Match(PyStr.Strip(line));
            if (opener is { } o)
            {
                flags.Add(true);
                if (m.Success && ClosesFence(m.Groups[1].Value, o)) opener = null;
                continue;
            }
            if (m.Success)
            {
                opener = (m.Groups[1].Value[0], m.Groups[1].Value.Length);
                flags.Add(true);
                continue;
            }
            flags.Add(false);
        }
        return flags;
    }

    public static string Slugify(string title)
    {
        var slug = PyStr.Strip(title).ToLowerInvariant();
        slug = SlugStripRe.Replace(slug, "");
        slug = SlugWhitespaceRe.Replace(slug, "-");
        return PyStr.Strip(slug, "-");
    }

    static string DedupeAnchor(string b, Dictionary<string, int> seen)
    {
        var count = seen.GetValueOrDefault(b, 0) + 1;
        seen[b] = count;
        return count == 1 ? b : $"{b}-{count}";
    }

    sealed record Section(string HeadingPath, string? Anchor, int StartLine, List<string> Lines);

    static List<Section> SplitIntoSections(string text)
    {
        var lines = PyStr.SplitLines(text);
        var stack = new List<(int Level, string Title, string Pinned)>();
        var sections = new List<Section>();
        var currentLines = new List<string>();
        int currentStart = 1;
        (char, int)? opener = null;
        var seenAnchors = new Dictionary<string, int>(StringComparer.Ordinal);

        void Flush()
        {
            var headingPath = string.Join(" > ", stack.Select(s => s.Title));
            string? anchor = null;
            if (stack.Count > 0)
            {
                var (_, title, pinned) = stack[^1];
                anchor = DedupeAnchor(pinned.Length > 0 ? pinned : Slugify(title), seenAnchors);
            }
            sections.Add(new Section(headingPath, anchor, currentStart, new List<string>(currentLines)));
        }

        for (int i = 0; i < lines.Count; i++)
        {
            int lineno = i + 1;
            var line = lines[i];
            var fence = FenceRe.Match(PyStr.Strip(line));
            if (opener is { } o)
            {
                currentLines.Add(line);
                if (fence.Success && ClosesFence(fence.Groups[1].Value, o)) opener = null;
                continue;
            }
            if (fence.Success)
            {
                opener = (fence.Groups[1].Value[0], fence.Groups[1].Value.Length);
                currentLines.Add(line);
                continue;
            }
            var heading = HeadingRe.Match(line);
            if (heading.Success)
            {
                Flush();
                int level = heading.Groups[1].Value.Length;
                var title = PyStr.Strip(heading.Groups[2].Value);
                var pinnedMatch = PinnedAnchorRe.Match(title);
                var pinned = pinnedMatch.Success ? pinnedMatch.Groups[1].Value : "";
                if (pinnedMatch.Success) title = PyStr.Strip(title.Substring(0, pinnedMatch.Index));
                while (stack.Count > 0 && stack[^1].Level >= level) stack.RemoveAt(stack.Count - 1);
                stack.Add((level, title, pinned));
                currentLines = [];
                currentStart = lineno + 1;
                continue;
            }
            currentLines.Add(line);
        }
        Flush();
        return sections.Where(s => PyStr.Strip(string.Join("\n", s.Lines)).Length > 0).ToList();
    }

    sealed record Block(int StartLine, string Text, bool IsFence);

    static List<Block> SplitIntoBlocks(List<string> lines, int startLine)
    {
        var blocks = new List<Block>();
        var cur = new List<string>();
        int curStart = startLine;
        (char, int)? opener = null;
        for (int offset = 0; offset < lines.Count; offset++)
        {
            int lineno = startLine + offset;
            var line = lines[offset];
            var stripped = PyStr.Strip(line);
            var fence = FenceRe.Match(stripped);
            if (opener is { } o)
            {
                cur.Add(line);
                if (fence.Success && ClosesFence(fence.Groups[1].Value, o))
                {
                    blocks.Add(new Block(curStart, string.Join("\n", cur), true));
                    cur = [];
                    opener = null;
                }
                continue;
            }
            if (fence.Success)
            {
                if (cur.Count > 0) { blocks.Add(new Block(curStart, string.Join("\n", cur), false)); cur = []; }
                curStart = lineno;
                cur = [line];
                opener = (fence.Groups[1].Value[0], fence.Groups[1].Value.Length);
                continue;
            }
            if (stripped.Length == 0)
            {
                if (cur.Count > 0) { blocks.Add(new Block(curStart, string.Join("\n", cur), false)); cur = []; }
                continue;
            }
            if (cur.Count == 0) curStart = lineno;
            cur.Add(line);
        }
        if (cur.Count > 0) blocks.Add(new Block(curStart, string.Join("\n", cur), opener is not null));
        return blocks;
    }

    static List<string> WrapByChars(string text, int maxChars)
    {
        var words = PyStr.SplitWhitespace(text);
        if (words.Count == 0) return [];
        var pieces = new List<string>();
        var cur = new List<string>();
        int curLen = 0;
        foreach (var word in words)
        {
            int extra = PyStr.Len(word) + (cur.Count > 0 ? 1 : 0);
            if (cur.Count > 0 && curLen + extra > maxChars)
            {
                pieces.Add(string.Join(" ", cur));
                cur = [word];
                curLen = PyStr.Len(word);
            }
            else
            {
                cur.Add(word);
                curLen += extra;
            }
        }
        if (cur.Count > 0) pieces.Add(string.Join(" ", cur));
        return pieces;
    }

    static List<(int, string)> PackUnits(List<(int Start, string Text)> units, int maxChars)
    {
        var packed = new List<(int, string)>();
        var curTexts = new List<string>();
        int curStart = 0, curLen = 0;
        foreach (var (start, text) in units)
        {
            int extra = PyStr.Len(text) + (curTexts.Count > 0 ? 2 : 0);
            if (curTexts.Count > 0 && curLen + extra > maxChars)
            {
                packed.Add((curStart, string.Join("\n\n", curTexts)));
                curTexts = [text];
                curStart = start;
                curLen = PyStr.Len(text);
            }
            else
            {
                if (curTexts.Count == 0) curStart = start;
                curTexts.Add(text);
                curLen += extra;
            }
        }
        if (curTexts.Count > 0) packed.Add((curStart, string.Join("\n\n", curTexts)));
        return packed;
    }

    public static List<Chunk> ChunkMarkdown(string text, int maxChars = 1200)
    {
        var chunks = new List<Chunk>();
        foreach (var section in SplitIntoSections(text))
        {
            var units = new List<(int, string)>();
            foreach (var block in SplitIntoBlocks(section.Lines, section.StartLine))
            {
                if (block.IsFence || PyStr.Len(block.Text) <= maxChars) units.Add((block.StartLine, block.Text));
                else units.AddRange(WrapByChars(block.Text, maxChars).Select(p => (block.StartLine, p)));
            }
            foreach (var (start, body) in PackUnits(units, maxChars))
                chunks.Add(new Chunk(section.HeadingPath, section.Anchor, start, body));
        }
        return chunks;
    }

    public static string EmbedText(Chunk chunk) => chunk.HeadingPath.Length == 0 ? chunk.Body : $"{chunk.HeadingPath}\n{chunk.Body}";
}
